src/Storage/FileStorage.php

```diff
@@ -1,476 +1,365 @@
-				StorageItem::METHOD_CLEAR, null, microtime(true) - $startTime));
-		}
-	}
-
-	/**
-	 * Returns if the backend is persistent or volatile.
-	 *
-	 * If the backend is volatile a system or service restart may destroy all the stored data.
-	 *
-	 * @return bool
-	 */
-	public function isPersistent() {
-		// File storage is always persistent.
-		return true;
-	}
-
-	/**
-	 * Returns whether the TTL functionality is supported by the backend.
-	 *
-	 * @return bool
-	 */
-	public function isTtlSupported() {
-		// If the storePlainText option is set to false, we support TTL functionality.
-		return !$this->storePlainText;
-	}
-
-	/**
-	 * Returns TRUE if the storage backend is read only, FALSE otherwise.
-	 *
-	 * @return bool
-	 */
-	public function isReadOnly() {
-		return $this->readOnly;
-	}
-
-	/**
-	 * Returns the configuration data for the storage backend. This includes the storage type as used by
-	 * the storage factory.
-	 *
-	 * @return array
-	 */
-	public function getConfigData() {
-		return array(
-			'storageType'      => StorageFactory::TYPE_FILE,
-			'path'             => $this->path,
-			'storePlainText'   => $this->storePlainText,
-			'filePrefix'       => $this->filePrefix,
-			'fileSuffix'       => $this->fileSuffix,
-			'fileMode'         => $this->fileMode,
-			'hashKey'          => $this->hashKey,
-			'readOnly'         => $this->readOnly,
-			'debuggerDisabled' => $this->debuggerDisabled,
-		);
-//			'path',
-//			'storePlainText',
-//			'filePrefix',
-//			'fileSuffix',
-//			'fileMode',
-//			'hashKey',
-//			'readOnly',
-//			'debuggerDisabled'
-
-	}
-}
\ No newline at end of file
+class FileStorage extends StorageAbstract
+{
+
+    /** @var string */
+    protected $path;
+
+    /** @var bool */
+    protected $canOnlyStorePlainText = false;
+
+    /** @var string */
+    protected $filenamePrefix = '';
+
+    /** @var string */
+    protected $filenameSuffix = '';
+
+    /** @var int */
+    protected $fileModeOctal = 0644;
+
+    /** @var bool */
+    protected $hashKey = false;
+
+    /** @var bool */
+    protected $readOnly;
+
+    /** @var FileHandlerPhp */
+    protected $fileHandler;
+
+    /**
+     * @throws ParameterException
+     * @throws StorageException
+     */
+    public function __construct(IFileHandler $fileHandler, string $path, bool $readOnly = false)
+    {
+        $this->fileHandler = $fileHandler;
+        $this->readOnly    = $readOnly;
+
+        $this->setPath($path);
+    }
+
+    /**
+     * @throws ParameterException
+     * @throws StorageException
+     */
+    public function setPath(string $path)
+    {
+        if (empty($path)) {
+            throw new ParameterException('Path has to be set');
+        }
+
+        if (substr($path, -1, 1) !== DIRECTORY_SEPARATOR) {
+            $path .= DIRECTORY_SEPARATOR;
+        }
+
+        $this->path = $path;
+
+        $this->initAndValidatePath();
+    }
+
+    /**
+     * @throws StorageException
```

Comment:
Would be nice to make specific exceptions to the different error types.

Reply:
yeah will need a lot more exceptions but you started the exception refact in the routing for some reason and don't want to conflict with anything there as it's not merged.
So at the end we'll need to go  through and check the exceptions